msg_queue.go

```diff
@@ -28,6 +32,27 @@ func (m *msgQueue) pushMessage(message *MessageReq) {
 	heap.Push(queue, message)
 }
 
+// pushPendingCommitMessage adds a new commit message to a message queue
+func (m *msgQueue) pushPendingCommitMessage(message *MessageReq) {
+	m.queueLock.Lock()
+	defer m.queueLock.Unlock()
+
+	m.pendingCommitMsgsQueue.Push(message)
+}
+
+// processPendingCommitMessages processes accumulated pending commit messages for the given sequence and round.
+func (m *msgQueue) processPendingCommitMessages(view *View, handler func(*MessageReq)) {
+	for {
+		m.queueLock.Lock()
+		msg, _ := m.pendingCommitMsgsQueue.readMessageWithDiscardsLocked(view, ValidateState)
+		m.queueLock.Unlock()
+		if msg == nil {
+			return
+		}
+		go handler(msg)
```

Comment:
What happens if:
- There are 10 items on the pending queue.
- After `AcceptState` they get unlocked and verified as part of `processPendingCommitMessages` (this is concurrent).
- The round changes and the old verification is still going. The proposal will be wrong.

I do not see this function being totally idempotent.

Reply:
I see that if the round changes, we would discard those messages, as they would be from the previous round(s). That's the main benefit and rationale for using `msgQueueImpl`. So I'd say we shouldn't be validating commit messages against the wrong proposal (i.e. the one being from not from the current sequence and round). 
However I want to verify it with additional logs.

Reply:
I think we have problem here. `PushMessage` used to be very simple, it used to just add new message to queue. With latest changes we introduced access to `view`, `proposal` and `velidatorset` from state inside method. And after that we have problem because those fields are accessed from more than one `go routine`.

Reply:
I agree with @igorcrevar. The logic is really convoluted. It is hard to keep a mental model of the lifecycle of a commit message.

Reply:
What @igorcrevar was trying to explain is that we would need to have lock when accessing the state reference. That is because we are accessing to the state from both `PushMessage`, which is executed in separate routine. We need to do that, because message queue without the pbft state isn't able to determine on its own whether given commit message is valid. We need proposal hash, current view (sequence and round) and validator set, which all are defined in the state.

Not sure if we can avoid locking the state even if we go with your approach, since we are going to be accessing to it from multiple routines eventually. 🤔